### PR TITLE
[WIP] [Scheduler] Add Etcd installation & modules

### PR DESCRIPTION
**Pull request by KeonHee**
## Description

This pr covers the [etcd](https://etcd.io/) installation and module implementation. And the etcd will be the coordinator of cluster members and the metadata store in the new scheduler.

We will implement modules based on the order defined in this page and add design documents under the page as well:
https://cwiki.apache.org/confluence/display/OPENWHISK/Component+Design

But I can not write wiki document because I do not have write access to the wiki :(
If you give me write access, I will write a document related to the etcd modules.

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [x] Data stores (e.g., CouchDB)
- [x] Tests
- [x] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [x] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

### CODE REVIEW

**Comment by mhenke1** [comment]
@KeonHee As discussed before, please put the new test code behind a feature toggle in a way that they are not executed as long the "new architecture" is not exploited (you could use the "assume" assertion of Scalatest based on a Ansible property).

**Comment by mhenke1** [comment]
@KeonHee  @style95  Was the usage of another "distributed state store" beside Zookeeper discussed on the mailing list or somewhere else ? On a first glance this feels wrong to add more complexity into the system.

**Comment by sven-lange-last** [comment]
My general comment on this PR is that it unconditionally adds system tests like `EtcdClientTests` that require a deployed etcd. In the dev mailing list discussion, a feature flag was requested for the architecture change: https://lists.apache.org/thread.html/63696c7bee9effb57118fe24513789d5538ea3a64af1204fccde73e8@%3Cdev.openwhisk.apache.org%3E.

By default, etcd should not be deployed and tests requiring etcd should not be run.

**Comment by style95** [comment]
@mhenke1 
Thank you for the feedback.
I have wanted this kind of discussion so that I shared my proposal via the mailing list and bi-weekly meeting.
https://lists.apache.org/list.html?dev@openwhisk.apache.org:lte=1M:Dominic%20Kim
https://www.youtube.com/watch?v=cPQ_UOEK9d8

Regarding etcd and zookeeper, I don't want to list up all the differences between them here because there are already many articles and materials out there.

I have used and loved zookeeper for many years.
But IMHO etcd is more lightweight and easy to operate than zookeeper.

And IIRC, we are using zookeeper only for Kafka.
Since my proposal is based on the assumption that we will rule out the kafka from the critical path in the future, so I believe we can replace zookeeper with etcd at some point.

@sven-lange-last 
Yes, we will definitely add switches to toggle our implementation and all modules will be disabled by default. Most of the modules are written in new files and based on SPI, I think we can easily achieve it.
Thank you for the good suggestion.

File: ansible/group_vars/all [hunk 1]
```diff
@@ -387,3 +387,26 @@ metrics:
...
+    port: 2480
+  cluster:
+    token: "{{ etcd_cluster_token | default('openwhisk-etcd-token') }}"
+  dir:
+    data: "{{ config_root_dir }}/etcd"
+  lease:
+    timeout: "{{ etcd_lease_timeout | default(1) }}"
+  quota_backend_bytes: "{{ etcd_quota_backend_bytes | default(0) }}"
+  snapshot_count: "{{ etcd_snapshot_count | default(100000) }}"
+  auto_compaction_retention: "{{ etcd_auto_compaction_retention | default(0) }}"
```
Comment (style95) [inline ansible/group_vars/all]: `retetion` should not be 0?

Comment (KeonHee) [reply ansible/group_vars/all]: `retention == 1` is more better. I will update it :)

**Comment by KeonHee** [comment]
@sven-lange-last @mhenke1 
Thank you for the feedback :)

Following yours feedback, I will create a test toggle.

**Comment by style95** [comment]
closing.
We will open a new one when the official procedure for proposals is defined.

**Pull request closed without merging**